### PR TITLE
Folders: Fix folder pagination for cloud instances with many folders

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

Filter k6 folder out during the SQL query rather than during post processing to ensure that the correct number of results is always returned and that frontend keeps fetching folders until the end of the folder list is reached.

**Why do we need this feature?**

Filtering out the k6 folder after the results were fetched meant that we're returning one less result to the frontend, which the frontend interprets as reaching the end of the folder list.

We want to keep hiding the k6 folder, as we don't want users to store dashboards or alerts in it, accidentally delete it etc.

**Who is this feature for?**

HG users

**Special notes for your reviewer:**

I've tested this with:
* users and service accounts (service accounts should keep seeing the k6 folder)
* with over 100 root level folders (to ensure that pagination works correctly)

### CODE REVIEW

**Review by DanCech — APPROVED** [review]
Looks reasonable

**Review by d14c — APPROVED** [review]

**Review by papagian — DISMISSED** [review]
I think that the query does not exclude nested folders under the excluded folder and dashboards living under them

File: pkg/services/sqlstore/searchstore/filters.go [hunk 1]
```diff
@@ -127,6 +128,14 @@ func (f DashboardFilter) Where() (string, []any) {
 	return sqlUIDin("dashboard.uid", f.UIDs)
 }
 
+type K6FolderFilter struct{}
+
+func (f K6FolderFilter) Where() (string, []any) {
+	filter := "dashboard.uid != ? AND (dashboard.folder_uid != ? OR dashboard.folder_uid IS NULL)"
```
[the remark this answers is not in the document]
Comment (d14c) [reply pkg/services/sqlstore/searchstore/filters.go:134]: @papagian as already discussed in slack, they should be only level of nesting in k6-app case, the addtional dashbaords have been used at some point as paceholders to alert the customers that the paremt folder is managed by k6-app and should not be deleted. Now that folders are hidden, those dushboards are not needed anymore. We have disabled provisioning dummy dashbaords in production and we will start deleting existing ones to remove any confusion.

[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/sqlstore/searchstore/filters.go:134]: Regarding org ID filter - it gets applied through a separate filter ([this one](https://github.com/grafana/grafana/blob/254610dce52073754d1cf52a6c0539f2f25f7b9b/pkg/services/sqlstore/searchstore/filters.go#L37-L43)), so I think that should be enough? We don't append org ID for any of the other UID filters (eg, `DashboardIDFilter` and `DashboardFilter`).

File: pkg/services/folder/folderimpl/sqlstore.go [hunk 2]
```diff
@@ -319,6 +321,13 @@ func (ss *sqlStore) GetChildren(ctx context.Context, q folder.GetChildrenQuery)
 			}
 			sql.WriteString(")")
 		}
+
+		// only list k6 folders when requested by a service account - prevents showing k6 folders in the UI for users
+		if q.SignedInUser == nil || q.SignedInUser.GetID().Namespace() != identity.NamespaceServiceAccount {
```
[the remark this answers is not in the document]
Comment (IevaVasiljeva) [reply pkg/services/folder/folderimpl/sqlstore.go:326]: Thanks for pointing it out! I'll keep this in mind for future PRs, and am also happy to rework this in the future if needed.

**Pull request merged**